homeassistant/helpers/config_validation.py

```diff
@@ -1063,6 +1063,7 @@ def script_action(value: Any) -> dict:
     vol.Optional(CONF_ALIAS): string,
     vol.Optional(CONF_CONTINUE_ON_ERROR): boolean,
     vol.Optional(CONF_ENABLED): boolean,
+    vol.Optional(CONF_VARIABLES): SCRIPT_VARIABLES_SCHEMA,
```

Comment:
I guess the one thing that we should be clear about is _when_ the variables apply. It should be consistently be applied after the step has run in which they are applied.

So applied once the trigger has triggered, the condition check has passed, the service has been executed or the wait for trigger triggered/timed out.

Reply:
Yeah, so the idea is to be clear on / the idea is:

Any variable applied to an action/condition/trigger only applies down the chain.

The fun part: The variable action applies for the script. But we could consider making that globally. Not sure on how to solve that part yet.

However, I think solving scoping all other levels first is important to be able to introduce global scoping (if we ever do that, that is a separate discussion).